File: homework4/4/stack.cpp
#include "stack.h"


Stack *createStack() {
    Stack *stack = new Stack{nullptr};
    return stack;
}

bool isEmpty(Stack *stack) {
    return stack->head == nullptr;
}

void push(Stack *stack, int value) {
    if (isEmpty(stack)) {
        stack->head = new StackElement{value,nullptr};
        return;
    }

    StackElement *tmp = stack->head;
    stack->head = new StackElement{value,tmp};
    return;
}
int pop(Stack *stack) {
    int result = -1;

    if (!isEmpty(stack)) {
        result = stack->head->value;
        StackElement *tmp = stack->head;
        stack->head = stack->head->next;
        delete tmp;
    }

    return result;
}

void deleteStack(Stack *stack) {
    while (!isEmpty(stack))
        pop(stack);
    delete stack;
}

File: java.1.1/Stack.java
package com.spbu.group144.Dementiev;

public class Stack {
    private StackElement head;

    private class StackElement {
        int value;
        StackElement next;
    }
    
    void push(int data) {
        StackElement newElement = new StackElement();
        newElement.value = data;

        if (head == null) {
            head = newElement;
        }
        else {
            newElement.next = head;
            head = newElement;
        }

    }

    int pop() {
        if (head == null) {
            System.out.println("Stack is empty");
            return 42;
        }

        if (head.next == null) {
            int data = head.value;
            head = null;
            return data;
        }

        int data = head.value;
        head = head.next;
        return data;
    }

    void printStack() {
        StackElement tmp = head;
        while (tmp != null) {
            System.out.println(tmp.value);
            tmp = tmp.next;
        }
    }
}
